[Spot] Add spot job name in the `SKYPILOT_TASK_ID` env var

Michaelvll opened the pull request:
<!-- Describe the changes in this PR -->

Fixes #3391

This PR adds the task name in the `SKYPILOT_TASK_ID`.

<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] `sky spot launch -n test "echo \$SKYPILOT_TASK_ID" --cpus 2`
    ```
    sky-managed-2024-04-05-17-47-48-461038_test_14-0
    ```
  - [x] `sky launch -c my-clus -n test "echo \$SKYPILOT_TASK_ID" --cpus 2`
    ```
    sky-2024-04-05-17-12-35-855707_my-clus_1
    ```
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

concretevitamin commented on docs/source/examples/spot-jobs.rst:
```diff
@@ -176,7 +176,7 @@ We also set :code:`--run_name` to :code:`$SKYPILOT_TASK_ID` so that the logs for
 
 .. note::
-  The environment variable :code:`$SKYPILOT_TASK_ID` (example: "sky-2022-10-06-05-17-09-750781_spot_id-22") can be used to identify the same job, i.e., it is kept identical across all
+  The environment variable :code:`$SKYPILOT_TASK_ID` (example: "sky-2022-10-06-05-17-09-750781_bert-qa_spot_id-22-1") can be used to identify the same job, i.e., it is kept identical across all
```
nit: Name in YAML is `bert_qa`. Should it be that or `bert-qa`?

Michaelvll replied:
Ahh, that is because we override the `-n` in the command line. I changed the `name:` to `bert-qa` in the task to reduce the confusion now : )

concretevitamin commented on docs/source/examples/spot-jobs.rst:
nit: We may want to mention manage spot's job ID in https://github.com/skypilot-org/skypilot/blob/3a52595cb58eb7e34189758c96069c431bb6e93d/docs/source/running-jobs/environment-variables.rst too.

Michaelvll replied:
Added. Thanks!

concretevitamin commented on docs/source/examples/spot-jobs.rst:
```diff
@@ -369,6 +369,12 @@ The above YAML file defines a pipeline with two tasks. The first :code:`name: pi
+.. note::
+
+  The :code:`$SKYPILOT_TASK_ID` environment variable is also available in the :code:`run` section of each task. It is unique for each task in the pipeline.
+  For the :code:`eval` task, the :code:`$SKYPILOT_TASK_ID` will be:
+  :code:`sky-2022-10-06-05-17-09-750781_pipeline_eval_spot_id-8-1`.
```
```suggestion
  For example, the :code:`$SKYPILOT_TASK_ID` for the :code:`eval` task above is:
  :code:`sky-2022-10-06-05-17-09-750781_pipeline_eval_spot_id-8-1`.
```

concretevitamin commented on docs/source/examples/spot-jobs.rst:
```diff
@@ -369,6 +369,12 @@ The above YAML file defines a pipeline with two tasks. The first :code:`name: pi
+.. note::
+
+  The :code:`$SKYPILOT_TASK_ID` environment variable is also available in the :code:`run` section of each task. It is unique for each task in the pipeline.
+  For the :code:`eval` task, the :code:`$SKYPILOT_TASK_ID` will be:
```
One UX discussion: I found myself reading these as "spot_id". Took a while to realize that it should be read as "_spot" and appended with "id-xxxx". Shall we skip the `_id` part if backward compatible?

Michaelvll replied:
Oh, it should be read as `spot_id` and appended with `_<job_id>-<task_id>`. I guess there will be no significant backward compatibility error in the case. I am not strongly for keeping the `id` or not. Wdyt?

concretevitamin replied:
Hmm, reason I found "spot_id" odd is that typically we can just append the IDs without saying IDs first. (E.g., we prepend the timestamp without saying `timestamp-`.)

In this case, how about we remove `-id`? I guess moving `spot-id` as a whole may be less desirable.

Michaelvll replied:
We do have an `_id` for normal cluster job. To keep it align, we may have to remove it for those jobs as well. 
Removed is for both now.

concretevitamin reviewed (commented):
Thanks @Michaelvll, some questions.

concretevitamin commented on docs/source/running-jobs/environment-variables.rst:
```diff
@@ -143,6 +143,8 @@ Environment variables for ``run``
        recoveries of that job will have the same ID value. Read more :ref:`here <spot-jobs-end-to-end>`.
      - sky-2023-07-06-21-18-31-563597_myclus_id-1
+     
+       managed spot job: sky-2023-07-06-21-18-31-563597_my-job-name_spot_id-1-1
```
```suggestion
       For managed spot jobs: sky-2023-07-06-21-18-31-563597_my-job-name_spot_id-1-1
```

Michaelvll commented:
I updated the naming of `$SKYPILOT_TASK_ID`, and added it for the setup section if the job is submitted with `--detach-setup` or `sky spot launch`. PTAL @concretevitamin

concretevitamin commented on docs/source/examples/spot-jobs.rst:
```diff
@@ -176,7 +176,7 @@ We also set :code:`--run_name` to :code:`$SKYPILOT_TASK_ID` so that the logs for
 
 .. note::
-  The environment variable :code:`$SKYPILOT_TASK_ID` (example: "sky-2022-10-06-05-17-09-750781_spot_id-22") can be used to identify the same job, i.e., it is kept identical across all
+  The environment variable :code:`$SKYPILOT_TASK_ID` (example: "sky-2022-10-06-05-17-09-750781_bert-qa_spot_id-22-1") can be used to identify the same job, i.e., it is kept identical across all
```
update?

concretevitamin commented on docs/source/running-jobs/environment-variables.rst:
```diff
@@ -108,12 +108,24 @@ Environment variables for ``setup``
    * - ``SKYPILOT_SETUP_NODE_IPS``
      - A string of IP addresses of the nodes in the cluster with the same order as the node ranks, where each line contains one IP address.
      - 1.2.3.4
+   * - ``SKYPILOT_TASK_ID``
+     - A unique ID assigned to each task.
+       
+       Please refer to the description in the :ref:`environment variables for run <env-vars-for-run>`.
+
+       This environment variable is available only when the task is submitted 
+       with :code:`sky launch --detach-setup`, or run as a managed spot job.
+     - sky-2023-07-06-21-18-31-563597_myclus_1
+     
+       managed spot job: sky-managed-2023-07-06-21-18-31-563597_my-job-name_1-1
```
```suggestion
   * - ``SKYPILOT_TASK_ID``
     - A unique ID assigned to each task.
       
       This environment variable is available only when the task is submitted 
       with :code:`sky launch --detach-setup`, or run as a managed spot job.
       
       Refer to the description in the :ref:`environment variables for run <env-vars-for-run>`.
     - sky-2023-07-06-21-18-31-563597_myclus_1
     
       For managed spot jobs: sky-managed-2023-07-06-21-18-31-563597_my-job-name_1-1
```
Should the last `-1` start with 0?

concretevitamin commented on docs/source/running-jobs/environment-variables.rst:
```diff
@@ -142,9 +154,9 @@ Environment variables for ``run``
        If a task is run as a :ref:`managed spot job <spot-jobs>`, then all
        recoveries of that job will have the same ID value. Read more :ref:`here <spot-jobs-end-to-end>`.
-     - sky-2023-07-06-21-18-31-563597_myclus_id-1
+     - sky-2023-07-06-21-18-31-563597_myclus_1
```
How about also including a logical schema for both cases in the cell above? `sky(-managed)-<timestamp>-<cluster_name>_<job_id>(-task_idx)` (please check correctness). This can help users parse it.

Does job_id start with 1 and task_idx 0? May need to update example values.

concretevitamin reviewed (approved):
LGTM, thanks @Michaelvll.
